flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -336,1697 +386,1982 @@ public int getKeyGroupPrefixBytes() {
-			}
+			FSDataInputStream inputStream = null;
+			FSDataOutputStream outputStream = null;
 
-			LOG.trace("Taking incremental snapshot for checkpoint {}. Snapshot is based on last completed checkpoint {} " +
-				"assuming the following (shared) files as base: {}.", checkpointId, lastCompletedCheckpoint, baseSstFiles);
+			try {
+				inputStream = remoteFileHandle.openInputStream();
+				stateBackend.cancelStreamRegistry.registerCloseable(inputStream);
 
-			// save meta data
-			for (Map.Entry<String, Tuple2<ColumnFamilyHandle, RegisteredKeyedBackendStateMetaInfo<?, ?>>> stateMetaInfoEntry
-				: stateBackend.kvStateInformation.entrySet()) {
-				stateMetaInfoSnapshots.add(stateMetaInfoEntry.getValue().f1.snapshot());
-			}
+				outputStream = restoreFileSystem.create(restoreFilePath, FileSystem.WriteMode.OVERWRITE);
+				stateBackend.cancelStreamRegistry.registerCloseable(outputStream);
 
-			// save state data
-			backupPath = new Path(stateBackend.instanceBasePath.getAbsolutePath(), "chk-" + checkpointId);
+				byte[] buffer = new byte[8 * 1024];
+				while (true) {
+					int numBytes = inputStream.read(buffer);
+					if (numBytes == -1) {
+						break;
+					}
 
-			LOG.trace("Local RocksDB checkpoint goes to backup path {}.", backupPath);
+					outputStream.write(buffer, 0, numBytes);
+				}
+			} finally {
+				if (stateBackend.cancelStreamRegistry.unregisterCloseable(inputStream)) {
+					inputStream.close();
+				}
 
-			backupFileSystem = backupPath.getFileSystem();
-			if (backupFileSystem.exists(backupPath)) {
-				throw new IllegalStateException("Unexpected existence of the backup directory.");
+				if (stateBackend.cancelStreamRegistry.unregisterCloseable(outputStream)) {
+					outputStream.close();
+				}
 			}
-
-			// create hard links of living files in the checkpoint path
-			Checkpoint checkpoint = Checkpoint.create(stateBackend.db);
-			checkpoint.createCheckpoint(backupPath.getPath());
 		}
 
-		KeyedStateHandle materializeSnapshot() throws Exception {
+		/**
+		 * In case of rescaling, this method creates a temporary RocksDB instance for a key-groups shard. All contents
+		 * from the temporary instance are copied into the real restore instance and then the temporary instance is
+		 * discarded.
+		 */
+		private void restoreKeyGroupsShardWithTemporaryHelperInstance(
+			Path restoreInstancePath,
+			List<ColumnFamilyDescriptor> columnFamilyDescriptors,
+			List<RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?>> stateMetaInfoSnapshots) throws Exception {
 
-			stateBackend.cancelStreamRegistry.registerCloseable(closeableRegistry);
+			List<ColumnFamilyHandle> columnFamilyHandles =
+				new ArrayList<>(1 + columnFamilyDescriptors.size());
 
-			// write meta data
-			metaStateHandle = materializeMetaData();
+			try (RocksDB restoreDb = stateBackend.openDB(
+				restoreInstancePath.getPath(),
+				columnFamilyDescriptors,
+				columnFamilyHandles)) {
```

Comment:
That might be true, but again this comment has nothing to do with this PR - only indentation was changed. I would strongly suggest that this should not become a collection for all possible suggestions to other changes. Opening a jira for this idea would be much better, because this thread is already very long and posting unrelated suggestions like this here make things more confusing and increases the likelihood that something will be overlooked. Would that be ok?

Reply:
Got it, ok!